### PR TITLE
cephadm: Allow users to provide ssh keys during bootstrap

### PR DESCRIPTION
**Pull request by ricardoasmarques**
With this PR user will be able to provide their own SSH keys:

```
ceph cephadm set-priv-key -i /root/.ssh/secret
ceph cephadm set-pub-key -i /root/.ssh/secret.pub
```

Even for the bootstrap:
```
cephadm boostrap \
    ...
    --ssh-private-key /root/.ssh/secret \
    --ssh-public-key /root/.ssh/secret.pub
```

Fixes: https://tracker.ceph.com/issues/45629

Signed-off-by: Ricardo Marques <rimarques@suse.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by ricardoasmarques** [comment]
@sebastian-philipp Can we assume that `cephadm` only works with keys that are not protected with passwords?

If yes, then we can have a single command `ceph cephadm set-key <private_key>` and extract the public key from the private key by doing something like `ssh-keygen -y -f <private_key>` internally.

File: src/pybind/mgr/cephadm/module.py [hunk 1]
```diff
@@ -747,6 +747,28 @@ def _generate_key(self):
             self._reconfig_ssh()
         return 0, '', ''
 
+    @orchestrator._cli_write_command(
+        'cephadm set-priv-key',
+        desc='Set cluster SSH private key (use -i <private_key>)')
+    def _set_priv_key(self, inbuf=None):
+        if inbuf is None or len(inbuf) == 0:
+            return -errno.EINVAL, "", "empty private ssh key provided"
+        self.set_store("ssh_identity_key", inbuf)
```
Comment (epuertat) [inline src/pybind/mgr/cephadm/module.py:756]: Is there any plan to remediate this? Storing private SSH keys in the clear (Is this somehow printed to the logs? Is this mon key-value store encrypted?) sounds like a serious attack vector: Dashboard's CVE-2020-1699 (or any other user-facing component breach) would expose the filesystem under `ceph` user, but with this data there everything else would be compromised, right?

Comment (ricardoasmarques) [reply src/pybind/mgr/cephadm/module.py:756]: @epuertat AFAIK, it's not possible to encrypt "config-keys". Note that this security issue is not being introduced in this PR, it's already present for the "generated SSH key" - https://github.com/ceph/ceph/blob/master/src/pybind/mgr/cephadm/module.py#L745

@sebastian-philipp was this security issue already discussed? Should we have a separate issue for that?

FYI, a similar discussion happened here https://github.com/ceph/ceph-iscsi/pull/173#discussion_r398688123

Comment (epuertat) [reply src/pybind/mgr/cephadm/module.py:756]: Thanks for the context, @ricardoasmarques!

File: src/cephadm/cephadm [hunk 2]
```diff
@@ -2562,31 +2562,39 @@ def command_bootstrap():
...
-        with open(auth_keys_file, 'a') as f:
-            os.fchmod(f.fileno(), 0o600)  # just in case we created it
-            if add_newline:
-                f.write('\n')
-            f.write(ssh_pub.strip() + '\n')
+        if args.ssh_private_key and args.ssh_public_key:
+            logger.info('Using provided ssh keys...')
+            mounts = {}
+            mounts[pathify(args.ssh_private_key)] = '/tmp/cephadm-ssh-key:z'
+            mounts[pathify(args.ssh_public_key)] = '/tmp/cephadm-ssh-key.pub:z'
```
[the remark this answers is not in the document]
Comment (ricardoasmarques) [reply src/cephadm/cephadm]: Done

File: src/cephadm/cephadm [hunk 3]
```diff
@@ -4464,6 +4472,13 @@ def _get_parser():
         '--dashboard-crt',
         help='Dashboard certificate')
 
+    parser_bootstrap.add_argument(
+        '--ssh-private-key',
+        help='SSH private key')
+    parser_bootstrap.add_argument(
+        '--ssh-public-key',
+        help='SSH public key')
```
[the remark this answers is not in the document]
Comment (ricardoasmarques) [reply src/cephadm/cephadm]: Done

**Comment by sebastian-philipp** [comment]
can we also set the ssh config?

File: src/pybind/mgr/cephadm/module.py (hunk 1 above)
Comment (sebastian-philipp) [reply src/pybind/mgr/cephadm/module.py:756]: I think, we need a proper and secure way to store secrets in the MONs. config-key and config are both not designed to store passwords etc. @jecluis any opinion? @travisn maybe something in the direction of k8s secrets might be an idea?

Comment (travisn) [reply src/pybind/mgr/cephadm/module.py:756]: @sebastian-philipp Rook certainly relies on K8s secrets, but not sure how cephadm could leverage something similar.

**Comment by ricardoasmarques** [comment]
> can we also set the ssh config?

@sebastian-philipp I've pushed a new commit that will alow users to provide ssh config during bootstrap (`--ssh-config` option).

**Comment by sebastian-philipp** [comment]
Big thanks! Do you also want to amend the man page? https://docs.ceph.com/docs/master/man/8/cephadm/

**Comment by ricardoasmarques** [comment]
> Big thanks! Do you also want to amend the man page? https://docs.ceph.com/docs/master/man/8/cephadm/

@sebastian-philipp Good catch. Man page is now updated.

**Comment by sebastian-philipp** [comment]
QA passed: http://pulpito.ceph.com/swagner-2020-05-27_13:56:17-rados:cephadm-wip-swagner-testing-2020-05-27-1352-distro-basic-smithi/

**Pull request merged**